controllers/kubeadmcontrolplane_controller.go

```diff
@@ -0,0 +1,202 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/utils/pointer"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/collections"
+	"sigs.k8s.io/cluster-api/util/patch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	anywherev1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/constants"
+)
+
+const (
+	kcpInPlaceUpgradeNeededAnnotation = "controlplane.clusters.x-k8s.io/in-place-upgrade-needed"
+	controlPlaneMachineLabel          = "cluster.x-k8s.io/control-plane-name"
+)
+
+// KubeadmControlPlaneReconciler reconciles a KubeadmControlPlaneReconciler object.
+type KubeadmControlPlaneReconciler struct {
+	client client.Client
+	log    logr.Logger
+}
+
+// NewKubeadmControlPlaneReconciler returns a new instance of KubeadmControlPlaneReconciler.
+func NewKubeadmControlPlaneReconciler(client client.Client) *KubeadmControlPlaneReconciler {
+	return &KubeadmControlPlaneReconciler{
+		client: client,
+		log:    ctrl.Log.WithName("KubeadmControlPlaneController"),
+	}
+}
+
+//+kubebuilder:rbac:groups=controlplane.cluster.x-k8s.io,resources=kubeadmcontrolplane,verbs=get;list;watch;update;patch
+//+kubebuilder:rbac:groups=controlplane.cluster.x-k8s.io,resources=kubeadmcontrolplane/status,verbs=get
+
+// Reconcile reconciles a KubeadmControlPlane object for in place upgrades.
+func (r *KubeadmControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, reterr error) {
+	log := r.log.WithValues("KubeadmControlPlane", req.NamespacedName)
+
+	log.Info("Reconciling KubeadmControlPlane object")
+	kcp := &controlplanev1.KubeadmControlPlane{}
+	if err := r.client.Get(ctx, req.NamespacedName, kcp); err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, err
+		}
+		return ctrl.Result{}, err
+	}
+
+	if !r.inPlaceUpgradeNeeded(kcp) {
+		log.Info("In place upgraded needed annotation not detected, nothing to do")
```

Comment:
Can we remove this log or add a higher verbosity to it? This is going to surface even on other providers besides tinkerbell which won't support in-place